rfcs/0000-buildpack-resource.md

```diff
@@ -0,0 +1,218 @@
+
+The ClusterStore being a monolithic specification of all available buildpacks leads to performance issues within kpack. If the list of available buildpacks within a ClusterStore is lengthy the reconciliation of a single ClusterStore can take considerable time. If a single buildpack within the ClusterStore is unavailable reconciliation will fail which will cause the entire contents of the ClusterStore to be unavailable.
+
+## Outcome:
+
+Allow Buildpacks within kpack to be managed independently of other buildpacks with their own unique lifecycles with a new ClusterBuildpack/Buildpack resource or directly within the (Cluster)Builder resource. This enables the the ClusterStore to be eventually removed.
+
+## Actions to take:
+
+Introduce a new ClusterBuildpack and Buildpack resource that provides a buildpack to be available within a Cluster or a singular namespace and be used in ClusterBuilders and Builders.
+ 
+Enable buildpacks to be used directly within (Cluster)Builders 
+
+##### Buildpack Resource
+
+The Buildpack resource will allow for an individual buildpack to be configured with a single buildpackage image reference and included in a Builder with other Buildpacks versioned and managed independently.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Buildpack
+metadata:
+  name: my-custom-app-buildpack-0.3.1
+  namespace: my-kpack-using-namespace
+spec:
+  image: gcr.io/my-org/my-custom-app-buildpack:0.3.1
+```
+
+A Buildpack resource should be available to be used within all Builders within its namespace. A Buildpack should not be useable by any ClusterBuilder. Optional credentials should be provided to the Buildpack through a `serviceAccountName` reference.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Buildpack
+metadata:
+  name: my-custom-app-buildpack-0.3.1
+  namespace: my-kpack-using-namespace
+spec:
+  image: gcr.io/my-org/my-custom-app-buildpack:0.3.1
+  serviceAccountName: my-service-account
+```
+
+The status of Buildpack resource should include the same metadata to the ClusterStore. This is to enable kpack to calculate the exact expected digest of the Builder image without needing to communicate with the registry. Additionally, this metadata may be used by external tooling such as the [kp cli](https://github.com/pivotal/kpack). The status must include the id, buildpack api, diffID, digest, size, homepage, supported stacks, and order for [meta-buildpacks](https://buildpacks.io/docs/concepts/components/buildpack/#meta-buildpack) of each buildpack within the configured buildpackage image. 
```

Comment:
> This is to enable kpack to calculate the exact expected digest of the Builder image without needing to communicate with the registry.

Are there any potential scenarios in which an image digest calculated by kpack might differ from the digest reported by a registry? If so, could scenarios which are either probable or damaging be mitigated?

Reply:
This behavior already exists in kpack so, we know that https://github.com/google/go-containerregistry can accurately calculate the digest of the builder image.